aten/src/ATen/ConjugateFallback.cpp

```diff
@@ -0,0 +1,42 @@
+#include <ATen/ConjugateFallback.h>
+#include <ATen/native/UnaryOps.h>
+
+namespace at {
+
+    TORCH_LIBRARY_IMPL(_, Conjugate, m) {
+        m.fallback(torch::CppFunction::makeFromBoxedFunction<&conjugateFallback>());
+    }
+
+   Tensor conj_materialize(const Tensor& self) {
+        // this function assumes that the tensor input has it's conjugate bit set
+        self.set_conjugate(false);
```

Comment:
This is not actually sound because it will affect all other references to the tensor; you need to allocate a new tensor (sharing storage/sizes/strides) and then set the conjugate bit to false and then pass it to conj. Or maybe we come up with an alternate way to call `self.conj()` that bypasses the conjugate flag and interprets the tensor in the "original" way (similar to how AutoNonVariableTypeMode works).

Reply:
One reason I'm a little hesitant about a NoConjMode is that it makes the invariants on the system a little more difficult to understand. With your current system, it's easy to see how conjugate bits relate to the Conjugate dispatch key: before running the Conjugate dispatch key, you may have tensors with conjugate bit set; after the dispatch key, nothing has conjugate bit set. With NoConjMode, you may selectively "let tensors with conjugate bit" through to the backend, and so invariants get a little harder to see.

That being said, it's a little hard for me to understand how to structure things in a reasonable way *without* letting the conjugate bit pass through to the backend. For example, matrix multiply with fused conjugation should definitely be implemented in the CPU/CUDA backend, so we definitely have to pass the conjugate bit all the way to matmul. So perhaps the invariant isn't worth preserving.

Reply:
It sounds like you're saying that we can lump backend kernels into two groups:
- conjugation-aware: a (relatively small number of) backend/op pairs that can take in a conjugate flag, and do conjugation along with the backend operation (multiplication on CPU/CUDA) in a performant way (something we want to take advantage of). We DO want to pass the conjugate bit pass through to the backend in this case.
- conjugation-unaware: kernels that require conjugation to be performed separately (and part of the questions is whether we want to explicit do the conjugation in the kernel, or move the logic somewhere else like I'm doing in this PR). It would be (maybe) preferable to not pass the conjugate bit through to the backend, to make the system easier to reason about.

Would it be reasonable to make the dispatcher aware of the specific op/backend pairs that can take in a conjugation flag, and know to only call out to my conjugate fallback if (a) the conjugate key is set on a tensor input, and (b) the particular op/backend pair currently being called is not in that first category?

I guess that would make this conjugateFallback kernel not really a "fallback" anymore.  More like a fallback "for everything except this whitelist of op/backend pairs", which also might become pretty hard to reason about.

Reply:
> It would be (maybe) preferable to not pass the conjugate bit through to the backend, to make the system easier to reason about.

Yes. The only trouble is if you "accidentally" pass along a tensor with the conjugate bit to the backend, and if there is no sanity checking in the back, you will end up with hilarious bugs where the result is the conjugate of the expected result.

> Would it be reasonable to make the dispatcher aware of the specific op/backend pairs that can take in a conjugation flag, and know to only call out to my conjugate fallback if (a) the conjugate key is set on a tensor input, and (b) the particular op/backend pair currently being called is not in that first category?

Yes. In fact, we can do this by registering fallthrough functions to operations that support conjugation bit directly. It is probably not worth splitting this into a per-backend concept; it is probably better being an all or nothing deal.

> More like a fallback "for everything except this whitelist of op/backend pairs", which also might become pretty hard to reason about.

I agree. Which is why @zdevito's suggestion to just call `materialize_conj` manually in all the kernels is appealing. After all, you *already* had to do some coding to make complex work, this is just one more thing.